lib/spack/spack/util/executable.py

```diff
@@ -246,6 +256,17 @@ def streamify(arg, mode):
                     "\nExit status %d when invoking command: %s"
                     % (proc.returncode, cmd_line_string),
                 )
+        except subprocess.TimeoutExpired as te:
+            proc.kill()
+            out, err = proc.communicate()
+            tty.warn(
+                f"Command {cmd_line_string} experienced a timeout after\
+{timeout}s running process {proc.pid}"
```

Comment:
Sure, I'll have just the exception raise with that string.

Reply:
Apologies, I put two requests in this and it looks like you addressed the first but not the second:

> Secondly, I think we should always fail on timeout (i.e. if there is a timeout and we got here, always raise an exception).

Or do you disagree?

Reply:
I thought it perhaps warranted more discussion. I actually was on the fence about whether the compiler detection should fail on timeout or not and thinking about it more now, I don't think we should be. If we always fail on timeout, we'd have to either add a try/catch or a hanging executable we attempt to determine a compiler version for would prevent other compilers from being detected without a modification to the PATH/some other filesystem location yielded by out compiler detection heuristic, which doesn't seem correct to me. 
In general I find their to be some functionality in being able to control timeout behavior without the need to add a try catch for a timeout exception in the calling context.

Reply:
>  If we always fail on timeout, we'd have to either add a try/catch or a hanging executable we attempt to determine a compiler version for would prevent other compilers from being detected without a modification to the PATH/some other filesystem location yielded by out compiler detection heuristic

Apologies, I did not follow this

My high-level assessment is:

* Timeouts are for executing functions we expect *not* to timeout
* Therefore, we should just terminate the spack instance when we *do* get a timeout

There are other potential uses for timeouts. AFAIK right now, the only place we set a timeout, we also say to fail, so we aren't using the don't-fail-on-timeout logic; a specific example might help me understand the purpose.

Reply:
Apologies for the lack of clarity, I hope this makes my perspective here a bit clearer:

> AFAIK right now, the only place we set a timeout, we also say to fail, so we aren't using the don't-fail-on-timeout logic

That is, at the moment, correct, however, what I had tried to convey in my previous comment was that I actually think we shouldn't fail during a timeout in compiler detection. I think timeouts, in addition to being used to prevent hanging in places we don't expect, can also be used to hedge our bets in places where we don't have a reasonable assurance that what we're executing _shouldn't_ hang. A good example of this is actually the motivating case for these changes, where we're executing arbitrary executables during compiler and as such don't really have the assurance that what we're running _shouldn't_ hang, just that if it does hang, we don't want to interpret it as a viable compiler for use by Spack. So in that case, rather than killing the entire compiler detection process, we should just timeout, fail to interpret that executable as a compiler and move on.

Reply:
> A good example of this is actually the motivating case for these changes, where we're executing arbitrary executables during compiler and as such don't really have the assurance that what we're running shouldn't hang,

Interesting: my opinion is we fell into this because our compiler regexes were too loose, so there's a definite path to remediation.

We may want to be more flexible than that though: while IMO we should resolve these cases for users, it may be wrong to entirely prevent them from making progress until we fix an issue like the regex being too loose.

I worry though about the possibly confusing mixture of `fail_on_error` (which already exists as an option) and `fail_on_timeout`: I think timeouts can be considered errors, so we should just allow users to specify `fail_on_error=False` if they don't want a timeout to raise an exception.

* There may be a desire to distinguish timeouts from other types of errors, but I'd want to hear a use case for that before we offered both.

Reply:
I think `fail_on_error` as a fail on timeout directive is a great idea, will implement. I can't think of a scenario where we'd only want to fail on timeout, so if that bridge is ever crossed it can be handled then.

Reply:
Amended. Setting `fail_on_error` to `False` in the compiler detection makes the `ignore_errors` tuple and the associated `ignore_version_errors` class attribute a bit facile. Do we want to remove that interface entirely?

Reply:
I want `fail_on_error` to be true for now, because we don't have special distinctions/handling for these errors. I think another PR could add clear warning messages to users to report these errors as bugs (and the benefit of continuing is that they are not necessarily hard-blocked while we address an error finding a compiler they may not care about).

At that point, I still think `ignore_version_errors` will be useful, because it is for indicating return codes that we would otherwise consider errors but actually are not (i.e. cases where the command is not considered to have failed and we shouldn't print any warning at all).

Reply:
> I want `fail_on_error` to be true for now

In that case, I think this is a use case for distinguishing timeouts from other errors. We shouldn't require users to modify their PATH (Or whatever other env path or filesystem location we use to determine potential compiler binaries) to avoid being hardblocked if they have a perfectly valid binary that we pick up as a compiler but happens to hang on user input or a license. Since we don't have special distinctions for the other errors, but we can/do have one for timeouts I think we should avoid being needlessly strict with this one in support of keeping those other specific errors handled, and IMO that's sufficient to differentiate the errors from the timeout. Perhaps if another PR refactors how those errors are handled, we can consolidate `fail_on_error` and `fail_on_timeout` at that time.
Keeping `fail_on_error` True wihtout special timeout handling will effectively make this timeout change a bit frivolous as the compiler detection will still crash Spack (although with more helpful error messages) and users will still be hard blocked at compiler detection or install time.

Reply:
> We shouldn't require users to modify their PATH (Or whatever other env path or filesystem location we use to determine potential compiler binaries) to avoid being hardblocked

I agree but I want to handle that in a future PR. 

This PR was content to fail right up until the last commit ([c030527](https://github.com/spack/spack/pull/39736/commits/c030527dda2f6b4745326c985296ba090cdf4dc1)) where it reversed that policy (what I'm trying to ascertain here is if you've changed your mind over the course of the review and decided this is critical).

> Keeping fail_on_error True without special timeout handling will effectively make this timeout change a bit frivolous as the compiler detection will still crash Spack (although with more helpful error messages)

Yes, that is what I want: i.e. I acknowledge it doesn't do much now, but am content for it to lay the groundwork for more-interesting handling. If you are inclined to resolve all of this here, then the task becomes defining the precisely correct behavior for each of the following cases:

1. timeout failure
2. failure that is not a timeout and not an explicitly `ignore_errors`
3. a failure that is explicitly in ignore errors

[3] is clear IMO. [1] and [2] are special case handling: my inclination is that a caller should specifically intercept this error when it wants to handle it and generate an error message.

```
compiler_version_logic():
  try:
    exe()
  except ExeError:
    # this could be a timeout or some unexpected return code
    tty.warn(Problem with the {exe_path} for compiler {x} - report this to spack maintainers)
```

what is missing from this PR in its current state is the informative warning

Reply:
> (what I'm trying to ascertain here is if you've changed your mind over the course of the review and decided this is critical).

In short, yes, that's exactly what happened.

> If you are inclined to resolve all of this here

I am so inclined.

> the task becomes defining the precisely correct behavior for each of the following cases:
> 3. a failure that is explicitly in ignore errors

As you said, this is fairly obvious, doesn't really need changing from current approach. 

> 2. failure that is not a timeout and not an explicitly ignore_errors

This IMO, also doesn't require special handling from the `Executable` class side of things. For compiler detection, I don't see a real reason to deviate from the current paradigm of failing on an error not in `ignore_errors` as I assume the failing on errors was a deliberate choice, unless that paradigm wasn't selected for any particular reason (or should be refactored) then yes, I like your solution (although these paths aren't guaranteed to actually be compilers, so perhaps the message should be something like "compiler candidate {exe} encountered failure {failure}. If this is a valid compiler, report to Spack maintainers...etc")

> 1. timeout failure

This is IMO the only really special case here, as it's not an execution derived error, it's one we're essentially manufacturing to establish guardrails. There are absolutely contexts where this is an actual error, i.e. if `Executable("cmake")("..")` timed out, that would clearly be a genuine error (either with the CMake or the timeout constraint) whereas the motivating case for this PR wouldn't be an error if the arbitrary executable were to time out but other errors may be genuine. Additionally, there can be a case where a timeout is unexpected but actual errors are.  The final consideration is the case where we encounter an expected timeout and still want the output of the command before the timeout, which we'd be unable to acquire in the event of a Exception. To me, that indicates the need for decoupling from `fail_on_error` and to avoid raising that Exception in some cases in order to obtain the output. 
If we don't want the API to expose both a fail on timeout and a fail on error argument (I agree wholeheartedly there), then I think some alternatives are:
* Add a `timeout_is_error` argument, this would remove the confusing names and indicate clearly how a timeout is to be handled.
* Refactor `ignore_errors` to include Python exceptions (or add a different argument responsible for this)
* Just always raise a timeout error if we encounter one (IMO this is the worst because it precludes us from obtaining the output)

Reply:
> The final consideration is the case where we encounter an expected timeout and still want the output of the command before the timeout, 

Why would we want that other than to report an exception? I think I'm saying the same thing as

> Same as suggestion directly above but we include the output as an attr of the Exception so calling context can access it if desirable.

but if that seems different, let me know.

My suggested approach: timeouts raise an error for `fail_on_error` (and the exception includes whatever output was generated up to the termination point).

We update the exception hierarchy like:

```
ProcessError
  Timeout
  BadReturnCode
```

So callers *can* choose to intercept one or both of them.

I think the compiler logic will want set `fail_on_error=True`, but catch both cases and print a warning (whatever message is stored in the exception). However, `compilers/__init__.detect_version` already handles this which means I think our work of generating warnings here is nearly done (right now the errors are reported as debug output in `compilers/__init__.find_compilers`).

Does that seem reasonable and address your concerns? I think part of this is that some of your concerns already happen to be addressed, but at a higher level.

Reply:
> I think part of this is that some of your concerns already happen to be addressed, but at a higher level.

My concerns aren't really with a specific usage of Executable, but are more general to how we model the `Timeout` conceptually.

> I think the compiler logic will want set fail_on_error=True

Yes, this is was only not the case if we wanted to prevent timeout errors from halting detection.

> I think our work of generating warnings here is nearly done (right now the errors are reported as debug output in compilers/__init__.find_compilers).

Right, I just need to add the except for the Timeout error in that method and the related warning output statement.

Reply:
The current approach of reporting the failures as debug statements is generally fine IMO, did you think that needed changing to warnings? I think the debug messages are best in that case as, since we're testing arbitrary files, it's less of a cause for concern if something has gone wrong.

I've pushed changes that just add the timeout specific exception to the `catch` options when then captures and reports the error exactly as the other types of errors are being reported.

Reply:
I could see timeouts being more problematic in the sense that the timeout is long (so that kind of failure has greater impact on user experience even when we get past it). For now though, keeping it debug sounds fine to me.

Reply:
That's a good point, although I think warning may still cause undue concern. Either way, I've fixed the style check, rebased and pushed.